src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java

```diff
@@ -0,0 +1,241 @@
+                                    RegularAndStaticColumns columns,
+                                    int nowInSec,
+                                    WriteContext ctx,
+                                    IndexTransaction.Type transactionType,
+                                    Memtable memtable)
+    {
+        final Set<Index.Indexer> indexers =
+                indices.stream().filter(indexSelector)
+                                .map(i -> i.indexerFor(key, columns, nowInSec, ctx, transactionType, memtable))
+                                .filter(Objects::nonNull)
+                                .collect(Collectors.toSet());
+
+        return indexers.isEmpty() ? null : new StorageAttachedIndex.IndexerAdapter()
+        {
+            @Override
+            public void insertRow(Row row)
+            {
+                forEach(indexer -> indexer.insertRow(row));
+            }
+
+            @Override
+            public void updateRow(Row oldRow, Row newRow)
+            {
+                forEach(indexer -> indexer.updateRow(oldRow, newRow));
+            }
+
+            private void forEach(Consumer<Index.Indexer> action)
+            {
+                indexers.forEach(action);
+            }
+        };
+    }
+
+    @Override
+    public StorageAttachedIndexQueryPlan queryPlanFor(RowFilter rowFilter)
+    {
+        return StorageAttachedIndexQueryPlan.create(baseCfs, queryMetrics, indices, rowFilter);
+    }
+
+    @Override
+    public SSTableFlushObserver getFlushObserver(Descriptor descriptor, LifecycleNewTracker tracker, TableMetadata tableMetadata)
+    {
+        return null;
+    }
+
+    @Override
+    public boolean handles(IndexTransaction.Type type)
+    {
+        // to skip CleanupGCTransaction and IndexGCTransaction
+        return type == IndexTransaction.Type.UPDATE;
+    }
+
+    @Override
+    public Set<Component> getComponents()
+    {
+        return getComponents(indices);
+    }
+
+    // TODO: Currently returns an empty set for in-memory only implementation
+    static Set<Component> getComponents(Collection<StorageAttachedIndex> indices)
+    {
+        return Collections.emptySet();
+    }
+
+    @Override
+    public void handleNotification(INotification notification, Object sender)
+    {
+        if (notification instanceof MemtableRenewedNotification)
+        {
+            indices.forEach(index -> index.getIndexContext().renewMemtable(((MemtableRenewedNotification) notification).renewed));
+        }
+        else if (notification instanceof MemtableDiscardedNotification)
+        {
+            indices.forEach(index -> index.getIndexContext().discardMemtable(((MemtableDiscardedNotification) notification).memtable));
+        }
+    }
+
+    /**
+     * @return count of queryable indexes
+     */
+    public int totalQueryableIndexCount()
+    {
+        return (int) indices.stream().filter(baseCfs.indexManager::isIndexQueryable).count();
```

Comment:
What if we just allowed this to return a `long`, and then changed to a `Gauge<Long>` in the metrics, which seems to be the only user?

Reply:
I would prefer to keep this as an int. The number of `indices` can only ever be an int because it is a list. I think it would be inconsistent to make the metric a long. As an aside, I find it odd that the stream api doesn't consider this and have `countInt` and `countLong` (not really those names but).

Reply:
We could use something like `checkedCast()` to be absolutely sure, but otherwise, WFM